File: frontend/src/api/aiSafety.ts
import apiClient from "./client";

export interface SecurityIncident {
  id: string;
  tenant_id: string;
  title: string;
  description: string | null;
  severity: "critical" | "high" | "medium" | "low";
  category:
    | "unauthorized_access"
    | "data_breach"
    | "malware"
    | "availability"
    | "policy_violation"
    | "other";
  status: "open" | "investigating" | "contained" | "resolved" | "closed";
  reported_by: string | null;
  assigned_to: string | null;
  affected_systems: string[] | null;
  timeline: Array<{
    at: string;
    action: string;
    user_id: string | null;
    details: string;
  }>;
  root_cause: string | null;
  remediation: string | null;
  resolved_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface MaintenanceWindow {
  id: string;
  name: string;
  description: string | null;
  starts_at: string;
  ends_at: string;
  recurrence: "once" | "weekly" | "monthly";
  affected_devices: string[] | null;
  block_ai_operations: boolean;
  block_bulk_jobs: boolean;
  is_active: boolean;
  created_at: string;
}

export interface ApprovalRequest {
  id: string;
  tenant_id: string;
  title: string;
  description: string | null;
  risk_level: "critical" | "high" | "medium" | "low";
  operation_context: Record<string, unknown> | null;
  requester_id: string | null;
  requester_note: string | null;
  first_approver_id: string | null;
  first_approved_at: string | null;
  second_approver_id: string | null;
  second_approved_at: string | null;
  rejection_reason: string | null;
  rejected_by: string | null;
  rejected_at: string | null;
  status: "pending_first" | "pending_second" | "approved" | "rejected" | "expired";
  requires_two_approvals: boolean;
  expires_at: string | null;
  created_at: string;
}

export interface ErasureRequest {
  id: string;
  tenant_id: string;
  target_user_email: string;
  reason: string | null;
  legal_basis: string | null;
  status: "pending" | "in_progress" | "completed" | "rejected";
  rejection_reason: string | null;
  affected_tables: string[] | null;
  audit_summary: Record<string, unknown> | null;
  approved_by: string | null;
  approved_at: string | null;
  completed_at: string | null;
  created_at: string;
}

const BASE = "/ai-safety";

export const aiSafetyApi = {
  // Maintenance windows
  listWindows: (activeOnly = false) =>
    apiClient.get<MaintenanceWindow[]>(`${BASE}/maintenance-windows`, { params: { active_only: activeOnly } }).then(r => r.data),
  createWindow: (data: Partial<MaintenanceWindow>) =>
    apiClient.post<MaintenanceWindow>(`${BASE}/maintenance-windows`, data).then(r => r.data),
  updateWindow: (id: string, data: Partial<MaintenanceWindow>) =>
    apiClient.patch<MaintenanceWindow>(`${BASE}/maintenance-windows/${id}`, data).then(r => r.data),
  deleteWindow: (id: string) => apiClient.delete(`${BASE}/maintenance-windows/${id}`),
  getActiveWindow: () =>
    apiClient.get<MaintenanceWindow | null>(`${BASE}/maintenance-windows/active`).then(r => r.data),

  // Approvals
  listApprovals: (status?: string) =>
    apiClient.get<ApprovalRequest[]>(`${BASE}/approvals`, { params: status ? { status } : {} }).then(r => r.data),
  createApproval: (data: {
    title: string; description?: string; risk_level: string;
    operation_context?: Record<string, unknown>; requester_note?: string; requires_two_approvals?: boolean;
  }) => apiClient.post<ApprovalRequest>(`${BASE}/approvals`, data).then(r => r.data),
  approve: (id: string) => apiClient.post<ApprovalRequest>(`${BASE}/approvals/${id}/approve`).then(r => r.data),
  reject: (id: string, reason: string) =>
    apiClient.post<ApprovalRequest>(`${BASE}/approvals/${id}/reject`, { reason }).then(r => r.data),

  // Erasure (LGPD)
  listErasure: (status?: string) =>
    apiClient.get<ErasureRequest[]>(`${BASE}/erasure`, { params: status ? { status } : {} }).then(r => r.data),
  createErasure: (data: { target_user_email: string; reason?: string; legal_basis?: string }) =>
    apiClient.post<ErasureRequest>(`${BASE}/erasure`, data).then(r => r.data),
  approveErasure: (id: string) => apiClient.post<ErasureRequest>(`${BASE}/erasure/${id}/approve`).then(r => r.data),
  executeErasure: (id: string) => apiClient.post<ErasureRequest>(`${BASE}/erasure/${id}/execute`).then(r => r.data),
  rejectErasure: (id: string, reason: string) =>
    apiClient.post<ErasureRequest>(`${BASE}/erasure/${id}/reject`, { reason }).then(r => r.data),

  // Security Incidents (SIRP)
  listIncidents: (params?: { status?: string; severity?: string }) =>
    apiClient
      .get<SecurityIncident[]>(`${BASE}/incidents`, { params })
      .then((r) => r.data),
  createIncident: (data: {
    title: string;
    description?: string;
    severity: string;
    category: string;
    affected_systems?: string[];
  }) => apiClient.post<SecurityIncident>(`${BASE}/incidents`, data).then((r) => r.data),
  updateIncident: (
    id: string,
    data: {
      status?: string;
      assigned_to?: string;
      root_cause?: string;
      remediation?: string;
      resolved_at?: string;
    }
  ) =>
    apiClient.patch<SecurityIncident>(`${BASE}/incidents/${id}`, data).then((r) => r.data),
  addTimeline: (id: string, action: string, details = "") =>
    apiClient
      .post<SecurityIncident>(`${BASE}/incidents/${id}/timeline`, { action, details })
      .then((r) => r.data),
};

File: backend/app/models/ai_safety.py
"""ORM models — F33 IA Safety & Governança."""
from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import Boolean, DateTime, ForeignKey, String, Text
from sqlalchemy.dialects.postgresql import ARRAY, JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.sql import func

from app.database import Base


class MaintenanceWindow(Base):
    __tablename__ = "maintenance_windows"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tenant_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    starts_at: Mapped[datetime] = mapped_column(DateTime, nullable=False)
    ends_at: Mapped[datetime] = mapped_column(DateTime, nullable=False)
    recurrence: Mapped[str] = mapped_column(String(20), default="once")
    recurrence_day: Mapped[int | None] = mapped_column(nullable=True)
    affected_devices: Mapped[list | None] = mapped_column(JSONB, nullable=True)
    block_ai_operations: Mapped[bool] = mapped_column(Boolean, default=True)
    block_bulk_jobs: Mapped[bool] = mapped_column(Boolean, default=True)
    created_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())


class ApprovalRequest(Base):
    __tablename__ = "approval_requests"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tenant_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    title: Mapped[str] = mapped_column(String(300), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    operation_context: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    risk_level: Mapped[str] = mapped_column(String(20), default="high")
    requester_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), nullable=True)
    requester_note: Mapped[str | None] = mapped_column(Text, nullable=True)
    first_approver_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), nullable=True)
    first_approved_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    second_approver_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), nullable=True)
    second_approved_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    rejection_reason: Mapped[str | None] = mapped_column(Text, nullable=True)
    rejected_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), nullable=True)
    rejected_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    status: Mapped[str] = mapped_column(String(20), default="pending_first")
    requires_two_approvals: Mapped[bool] = mapped_column(Boolean, default=True)
    expires_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())


class ErasureRequest(Base):
    __tablename__ = "erasure_requests"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tenant_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    requested_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), nullable=True)
    target_user_email: Mapped[str] = mapped_column(String(255), nullable=False)
    reason: Mapped[str | None] = mapped_column(Text, nullable=True)
    legal_basis: Mapped[str | None] = mapped_column(String(100), nullable=True)
    status: Mapped[str] = mapped_column(String(20), default="pending")
    rejection_reason: Mapped[str | None] = mapped_column(Text, nullable=True)
    affected_tables: Mapped[list | None] = mapped_column(JSONB, nullable=True)
    audit_summary: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    approved_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), nullable=True)
    approved_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())


class SecurityIncident(Base):
    """SIRP — Registro de incidentes de segurança com timeline JSONB."""
    __tablename__ = "security_incidents"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tenant_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("tenants.id", ondelete="CASCADE"), nullable=False)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    severity: Mapped[str] = mapped_column(String(20), default="medium")
    category: Mapped[str] = mapped_column(String(50), default="other")
    status: Mapped[str] = mapped_column(String(20), default="open")
    reported_by: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    assigned_to: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    affected_systems: Mapped[list | None] = mapped_column(ARRAY(Text), nullable=True)
    timeline: Mapped[list] = mapped_column(JSONB, default=list)
    root_cause: Mapped[str | None] = mapped_column(Text, nullable=True)
    remediation: Mapped[str | None] = mapped_column(Text, nullable=True)
    resolved_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
